src/query/atomWithQuery.ts

```diff
@@ -28,76 +28,60 @@ export function atomWithQuery<
   equalityFn: (a: TData, b: TData) => boolean = Object.is
 ): WritableAtom<TData, Action> {
-  const stateAtom = atom(() => {
-    const state = {
-      resolve: null as ((data: TData) => void) | null,
-      setData: null as ((data: TData) => void) | null,
-      prevData: null as TData | null,
-      unsubscribe: null as (() => void) | null,
-      handle(result: QueryObserverResult<TData, TError>) {
+  const queryDataAtom = atom(
+    (get) => {
+      const queryClient = get(queryClientAtom)
+      const options =
+        typeof createQuery === 'function' ? createQuery(get) : createQuery
+      let resolve: ((data: TData) => void) | null = null
+      const dataAtom = atom<TData | Promise<TData>>(
+        new Promise<TData>((r) => {
+          resolve = r
+        })
+      )
+      let setData: (data: TData) => void = () => {
+        throw new Error('atomWithQuery: setting data without mount')
+      }
+      let prevData: TData | null = null
+      const listener = (result: QueryObserverResult<TData, TError>) => {
         // TODO error handling
         if (
           result.data === undefined ||
-          (this.prevData !== null && equalityFn(this.prevData, result.data))
+          (prevData !== null && equalityFn(prevData, result.data))
         ) {
           return
         }
-        this.prevData = result.data
-        if (this.resolve) {
-          this.resolve(result.data)
-          this.resolve = null
-        } else if (this.setData) {
-          this.setData(result.data)
+        prevData = result.data
+        if (resolve) {
+          resolve(result.data)
+          resolve = null
         } else {
-          throw new Error('setting data without mount')
+          setData(result.data)
         }
-      },
-    }
-    return state
-  })
-  const initAtom = atom(
-    (get) => {
-      const options =
-        typeof createQuery === 'function' ? createQuery(get) : createQuery
-      const state = get(stateAtom)
-      const dataAtom = atom<TData | Promise<TData>>(
-        new Promise<TData>((resolve) => {
-          state.resolve = resolve
+      }
+      const defaultedOptions = queryClient.defaultQueryObserverOptions(options)
+      if (typeof defaultedOptions.staleTime !== 'number') {
+        defaultedOptions.staleTime = 1000
+      }
+      const observer = new QueryObserver(queryClient, defaultedOptions)
+      observer
+        .fetchOptimistic(defaultedOptions)
+        .then(listener)
+        .catch(() => {
+          // TODO error handling
         })
-      )
-      state.prevData = null
-      const queryClient = get(queryClientAtom)
-      const observer = new QueryObserver(queryClient, options)
-      state.unsubscribe?.()
-      state.unsubscribe = observer.subscribe((result) => {
-        state.handle(result)
-      })
+      dataAtom.onMount = (update) => {
+        setData = update
+        const unsubscribe = observer.subscribe(listener)
```

Comment:
Can this throw? If so, it needs error handling.

Reply:
Looking at RQ code, it looks okay.
https://github.com/tannerlinsley/react-query/blob/9aed040364796cdd5cab0f53123e02a6c6f769de/src/react/useBaseQuery.ts#L85